dory/Hardware_targets/GAP8/Common/Utils/mchan.h

```diff
@@ -0,0 +1,139 @@
+#ifndef _MCHAN_H
+#define _MCHAN_H
+
+// Requires to have MCHAN_BASE_ADDR, MCHAN_EVENT defined outside of header
+#ifndef MCHAN_BASE_ADDR
+#error "[mchan.h] MCHAN_BASE_ADDR not defined!"
+#endif
+
+#if !defined(MCHAN_EVENT) && !defined(MCHAN_POLLED)
+#error "[mchan.h] Nor MCHAN_EVENT nor MCHAN_POLLED defined!"
+#endif
+
+#if defined(MCHAN_EVENT) && !defined(MCHAN_EVENT_BIT)
+#error "[mchan.h] MCHAN_EVENT_BIT should be defined when using events as signalization!"
+#endif
+
+#include "pmsis.h"
+
+#define MCHAN_CMD_OFFSET 0
+#define MCHAN_STATUS_OFFSET 4
+
+#define MCHAN_CMD_ADDR (MCHAN_BASE_ADDR + MCHAN_CMD_OFFSET)
+#define MCHAN_STATUS_ADDR (MCHAN_BASE_ADDR + MCHAN_STATUS_OFFSET)
+
+#define READ_REG(addr) (*(volatile int*)(addr))
+#define WRITE_REG(addr, value) do { *(volatile int*)(addr) = (int)value; } while (0)
+
+#define MCHAN_READ_CMD() READ_REG(MCHAN_CMD_ADDR)
+#define MCHAN_WRITE_CMD(value) WRITE_REG(MCHAN_CMD_ADDR, value)
+
+#define MCHAN_READ_STATUS() READ_REG(MCHAN_STATUS_ADDR)
+#define MCHAN_WRITE_STATUS(value) WRITE_REG(MCHAN_STATUS_ADDR, value)
+
+// MCHAN version 7 has 1 more bit for the transfer length, so all the flag offsets are shifted by 1. Also, LOC (TCDM) striding is not supported in v6.
+#if MCHAN_VERSION==7
+#define MCHAN_CMD_FLAG_DIRECTION_LOC2EXT    (0 << 17)
+#define MCHAN_CMD_FLAG_DIRECTION_EXT2LOC    (1 << 17)
+#define MCHAN_CMD_FLAG_INCREMENTAL          (1 << 18)
+#define MCHAN_CMD_FLAG_2D_TRANSFER_EXTERNAL (1 << 19)
+#define MCHAN_CMD_FLAG_EVENT_ENABLE         (1 << 20)
+#define MCHAN_CMD_FLAG_INTERRUPT_ENABLE     (1 << 21)
+#define MCHAN_CMD_FLAG_BROADCAST_FINISH     (1 << 22)
+#define MCHAN_CMD_FLAG_2D_TRANSFER_LOCAL    (1 << 23)
+#define MCHAN_CMD_SHIFT_DIRECTION (17)
+#else
+#define MCHAN_CMD_FLAG_DIRECTION_LOC2EXT    (0 << 16)
+#define MCHAN_CMD_FLAG_DIRECTION_EXT2LOC    (1 << 16)
+#define MCHAN_CMD_FLAG_INCREMENTAL          (1 << 17)
+#define MCHAN_CMD_FLAG_2D_TRANSFER_EXTERNAL (1 << 18)
+#define MCHAN_CMD_FLAG_EVENT_ENABLE         (1 << 19)
+#define MCHAN_CMD_FLAG_INTERRUPT_ENABLE     (1 << 20)
+#define MCHAN_CMD_FLAG_BROADCAST_FINISH     (1 << 21)
+#define MCHAN_CMD_SHIFT_DIRECTION (16)
+#endif
```

Comment:
I would write this maybe like this:

```suggestion
// MCHAN version 7 has 1 more bit for the transfer length, so all the flag offsets are shifted by 1. Also, LOC (TCDM) striding is not supported in v6.
#if MCHAN_VERSION==7
#define MCHAN_CMD_TRANSFER_LENGTH_SIZE (17)
#else
#define MCHAN_CMD_TRANSFER_LENGTH_SIZE (16)
#endif

#define MCHAN_CMD_FLAG_DIRECTION_LOC2EXT    (0 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 0))
#define MCHAN_CMD_FLAG_DIRECTION_EXT2LOC    (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 0))
#define MCHAN_CMD_FLAG_INCREMENTAL          (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 1))
#define MCHAN_CMD_FLAG_2D_TRANSFER_EXTERNAL (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 2))
#define MCHAN_CMD_FLAG_EVENT_ENABLE         (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 3))
#define MCHAN_CMD_FLAG_INTERRUPT_ENABLE     (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 4))
#define MCHAN_CMD_FLAG_BROADCAST_FINISH     (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 5))
#if MCHAN_VERSION == 7
#define MCHAN_CMD_FLAG_2D_TRANSFER_LOCAL    (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 6))
#endif
#define MCHAN_CMD_SHIFT_DIRECTION (MCHAN_CMD_TRANSFER_LENGTH_SIZE)
```

Reply:
@da-gazzi I think you forgot to add the flag for the local 2D transfer
```
#if MCHAN_VERSION == 7
#define MCHAN_CMD_FLAG_2D_TRANSFER_LOCAL    (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 6))
#endif
```

Reply:
both fixed!